server/autotest_server.py

```diff
@@ -679,3 +685,147 @@ def update_test_scripts(files_path, assignment_id, markus_address):
+    return h.digest()
+
+def deep_dict_equal(a, b, ignore=set()):
+    """
+    Return True iff dictionaries a and b are equal. This comparison does 
+    not take into account any keys in the ignore set or their values. Keys
+    at any nested level of dictionaries a and b will be ignored if they are
+    in the ignore set 
+    """
+    if isinstance(a, dict) and isinstance(b, dict):
+        ak = a.keys() - ignore
+        bk = b.keys() - ignore
+        if ak == bk:
+            return all(deep_dict_equal(a[k], b[k], ignore) for k in ak)
+        return False
+    try:
+        return all(deep_dict_equal(*z, ignore) for z in zip_longest(a,b) if z != (a, b))
+    except TypeError:
+        return a == b
+
+def settings_changed(env_dir, env_settings, checksum, specs_dir):
+    """
+    Return True iff the environment setting have changed or if the files
+    have been updated. This is calculated by comparing the old settings
+    found in env_dir to the new settings in env_settings and checking if the
+    checksum for the new files matches the old checksum. 
+    """
+    settings_file = os.path.join(env_dir, 'environment_settings.json')
+    checksum_file = os.path.join(env_dir, 'checksum')
+    ignore_file = os.path.join(specs_dir, 'evironment_ignore.yml')
+    if os.path.isfile(settings_file):
+        if os.path.isfile(checksum_file) and checksum is not None:
+            with open(checksum_file, 'rb') as f:
+                old_checksum = f.read()
+            return checksum == old_checksum
+        if os.path.isfile(ignore_file):
+            with open(ignore_file) as f:
+                ignore_keys = set(yaml.load(f))
+        else:
+            ignore_keys = set()
+        with open(settings_file) as f:
+            current_settings = json.load(f)
+        return deep_dict_equal(env_settings, current_settings, ignore=ignore_keys) 
+    return False
+
+@clean_after
+def manage_tester_environment(markus_address, tester_type, tester_name, env_settings, files_path):
+    """
+    Create a new tester environment or update an old one if it exists.  Errors are written to a file
+    called env_creation_errors.txt and are checked and reported the next time a test is run.
+
+    This function should be used by an rq worker.
+    """
+    env_name = get_unique_env_name(markus_address, tester_type, tester_name)
+    env_dir = os.path.join(config.WORKSPACE_DIR, config.SPECS_DIR_NAME, env_name)
+    error_file = os.path.join(env_dir, 'env_creation_errors.txt')
+    try:
+        # get the specs and bin directories for the given tester_type
+        tester_dir = get_tester_root_dir(tester_type)
+        specs_dir = os.path.join(tester_dir, 'specs')
+        bin_dir = os.path.join(tester_dir, 'bin')
+        # calculate a checksum for the uploaded files
+        if files_path is not None:
+            files = (fname for fd, fname in recursive_iglob(files_path) if fd == 'f')
+            checksum = generate_checksum(files)
```

Comment:
Have you tried how expensive is to do this for, say, 30 files?
(the usual sql solutions + datasets)

Reply:
I ran it on the contents of the old client directories. It's not instantaneous but I don't think it's going to be a real bottleneck. This only gets run once when the test scripts are updated, not during the actual test jobs.